### PR TITLE
8321176: [Screencast] make a second attempt on screencast failure

### PR DESCRIPTION
**Pull request by antbob**
This patch adds re-try logic to libpipewire screencast error handling as discussed in PR #16794 and also brings some additional error handling and thread safety improvements. Specifically around cleanup order where incorrect ordering lead to native memory corruption issues, and lock/unlock accounting that while mostly harmless (with the current libpipewire implementation) did pollute the stderr on jtreg tests, making some tests (expecting clean stderr) to fail.

The real major change here however is the throw of the RuntimeException which can propagate to public 

java.awt.Robot. createMultiResolutionScreenCapture, createScreenCapture, getPixelColor methods. I'm not sure the plain RuntimeException is the way to go here so this is just a placeholder of sorts. A separate/specific runtime exception can be created for this BUT *something* needs to be done here as the current implementation fails to convey libpipewire failures to those public API callers and since they have no way of detecting such errors otherwise they have no way of knowing that the data returned by those API is in fact invalid (eg black screens etc). The reason for using an unchecked exception here is driven mainly by the following factors:

1) Since those are public API, their contracts can potentially make it difficult to introduce specific additional checked exceptions or return values (as appropriate) as those could potentially break existing API use.

2) The libpipewire errors of that kind are rare and usually indicate there is something wrong with the desktop stack eg some fatal configuration or run time error that is normally not supposed to happen and given this patch now goes extra step re-trying on such failures it stands to reason runtime unchecked exception makes sense when that fails as well.

3) Creating checked exceptions for such specific native implementation dependent errors and propagating such exceptions thru the call tree does not make much sense as most public API users won't even know how to handle them without knowing native implementation specifics.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8321176](https://bugs.openjdk.org/browse/JDK-8321176): [Screencast] make a second attempt on screencast failure (**Bug** - P3)

### Reviewers
 * [Alexander Zvegintsev](https://openjdk.org/census#azvegint) (@azvegint - **Reviewer**)
 * [Phil Race](https://openjdk.org/census#prr) (@prrace - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16978/head:pull/16978` \
`$ git checkout pull/16978`

Update a local copy of the PR: \
`$ git checkout pull/16978` \
`$ git pull https://git.openjdk.org/jdk.git pull/16978/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16978`

View PR using the GUI difftool: \
`$ git pr show -t 16978`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16978.diff">https://git.openjdk.org/jdk/pull/16978.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16978#issuecomment-1841225622)

### CODE REVIEW

**Comment by antbob** [comment]
@azvegint please have a look and let me know what you think. i dont have access to the issue tracker so lets have a discussion here and adopt this patch accordingly if possible.

File: src/java.desktop/unix/classes/sun/awt/screencast/ScreencastHelper.java
Comment (prrace) [inline src/java.desktop/unix/classes/sun/awt/screencast/ScreencastHelper.java:223]: Please remove this

Comment (antbob) [reply src/java.desktop/unix/classes/sun/awt/screencast/ScreencastHelper.java:223]: @prrace have you read my description of this PR? If so, and based on that, what can you propose to address the problem described?

Comment (azvegint) [reply src/java.desktop/unix/classes/sun/awt/screencast/ScreencastHelper.java:223]: Such change requires a [CSR](https://wiki.openjdk.org/display/csr) .
If we want to get this fix into JDK 22, we don't have time for CSR, since JDK 22 will be forked this Thursday.

**Comment by antbob** [comment]
@azvegint thanks for review! yeah the CSR looks like a long story so lets skip it in this patch however i feel that the problem has to be addressed somehow going forward as i dont think the current behavior is acceptable. there could be cases where it can cause real problems eg imagine some sort of medical imaging app where it could fail intermittently and report wrong pixel colors etc. can you please file a bug based on my description above ?

and also something i thought i should mention wrt doCleanup() function. it might be a good idea to protect it with a mutex or something. it does look safe ATM but i haven't looked at related libpipewire internals to assert that with full certainty. there perhaps could be a situation where doCleanup() is called internally but also from java cleanup thread (the one that sleeps for 2 seconds) and if they interleave/race it might cause unexpected behavior or crash in libpipewire as its rather sensible in terms of teardown sequence (i've seen its crashing due to that in my testing and thats why i have reshuffled some related calls in this patch). anyways, this is something to look into and check.

**Review by azvegint — APPROVED** [review]
Looks good to me.

> can you please file a bug based on my description above ?

Filed the [JDK-8321475](https://bugs.openjdk.org/browse/JDK-8321475) for this

> and also something i thought i should mention wrt doCleanup() function. it might be a good idea to protect it with a mutex or something. it does look safe ATM but i haven't looked at related libpipewire internals to assert that with full certainty. there perhaps could be a situation where doCleanup() is called internally but also from java cleanup thread (the one that sleeps for 2 seconds) and if they interleave/race it might cause unexpected behavior or crash in libpipewire as its rather sensible in terms of teardown sequence (i've seen its crashing due to that in my testing and thats why i have reshuffled some related calls in this patch). anyways, this is something to look into and check.

Sure, we should revisit it later.

**Comment by antbob** [comment]
@azvegint super, thanks again!

File: src/java.desktop/unix/classes/sun/awt/screencast/ScreencastHelper.java
Comment (prrace) [reply src/java.desktop/unix/classes/sun/awt/screencast/ScreencastHelper.java]: It is beyond that. It is an incompatible change. It will be rejected forever.

File: src/java.desktop/unix/native/libawt_xawt/awt/screencast_pipewire.c
Comment (prrace) [inline src/java.desktop/unix/native/libawt_xawt/awt/screencast_pipewire.c:102]: I had this disconnect inside the lock in an earlier PR 
https://github.com/openjdk/jdk/pull/14428/files?diff=unified&w=0#diff-3ba5df79cdd3da36b21bf29b4e8de462dd61e6a21dfe0816e4d84c18bbfb76b2R92
but then it was moved out of the lock in another PR to fix a bug https://bugs.openjdk.org/browse/JDK-8310334
https://github.com/openjdk/jdk/pull/15250/files#diff-3ba5df79cdd3da36b21bf29b4e8de462dd61e6a21dfe0816e4d84c18bbfb76b2


So how can it be right to put it back inside the lock ?

File: src/java.desktop/unix/classes/sun/awt/screencast/ScreencastHelper.java
Comment (antbob) [reply src/java.desktop/unix/classes/sun/awt/screencast/ScreencastHelper.java]: @prrace the related API already throw SecurityException if capture permission is not granted. moreover the API has been changed already in the pastr to accommodate that eg see

 https://docs.oracle.com/javase/8/docs/api/java/awt/Robot.html#getPixelColor-int-int-

VS

https://docs.oracle.com/en/java/javase/21/docs/api/java.desktop/java/awt/Robot.html#getPixelColor(int,int)

and so it stands to reason that a similar solution can be found for this problem as well if there is a will to do so.

An exception seem to be the most logical solution here to me however it could be something different, like the approach you describe in JDK-8321475 (i cant comment there so commenting here instead). Either way, like i said, i firmly believe that the current API behavior is just plain wrong and potentially harmful.

File: src/java.desktop/unix/native/libawt_xawt/awt/screencast_pipewire.c
Comment (antbob) [reply src/java.desktop/unix/native/libawt_xawt/awt/screencast_pipewire.c:102]: @prrace the related previous fixes didnt really fix anything and just re-shuffled things around to dodge specific problems encountered. the libpipewire API locking and signaling has been fundamentally broken before my patch for JDK-8320655. this is understandable given libpipewire lack of proper documentation in that area and no clear specs on when and how locking and signaling should be done.

this particular case you refer to you need to lock around disconnect bc if you don't there is a chance the disconnect API can fail here

https://github.com/PipeWire/pipewire/blob/master/src/pipewire/thread-loop.c#L100

since this is recursive lock, taking it explicitly before calling disconnect ensures this does not happen. if you test the old implementation with libpipewire debug enabled you would see it constantly failing in related API and complaining functions being called in the "wrong context", including this code in question.

**Comment by antbob** [comment]
@azvegint @prrace unless there are any outstanding issues with this patch that need addressing/correcting can you please sponsor it for me?

**Comment by azvegint** [comment]
> @azvegint @prrace unless there are any outstanding issues with this patch that need addressing/correcting can you please sponsor it for me?

In the client, we have a rule that PR requires two people approval for integration (despite the hard rule of 1 mentioned in the header). 
So I am waiting for another approval.

**Review by prrace — APPROVED** [review]

**Pull request closed without merging**
